core/string_builder.h

```diff
@@ -0,0 +1,81 @@
+/*************************************************************************/
+/*  string_builder.h                                                     */
+/*************************************************************************/
+/*                       This file is part of:                           */
+/*                           GODOT ENGINE                                */
+/*                      https://godotengine.org                          */
+/*************************************************************************/
+/* Copyright (c) 2007-2017 Juan Linietsky, Ariel Manzur.                 */
+/* Copyright (c) 2014-2017 Godot Engine contributors (cf. AUTHORS.md)    */
+/*                                                                       */
+/* Permission is hereby granted, free of charge, to any person obtaining */
+/* a copy of this software and associated documentation files (the       */
+/* "Software"), to deal in the Software without restriction, including   */
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+#ifndef STRING_BUILDER_H
+#define STRING_BUILDER_H
+
+#include "core/reference.h"
+
+#include "core/ustring.h"
+
+#define BUCKET_SIZE 32
+
+class StringBuilder : public Reference {
+	GDCLASS(StringBuilder, Reference)
+private:
+	template <class T>
+	struct BuilderBucket {
+		T elements[BUCKET_SIZE];
+		uint32_t elements_used = 0;
+	};
+
+	uint32_t string_length = 0;
+
```

Comment:
This is probably too complex and both String and const char * are really small (both 8 bytes), and have no initialization/freeing cost if empty, so i would just do:
```

struct StringData {
   String string;
   cosnt char* cstring; //if not null, use this, otherwise use string.
}; // should be at most 16 bytes, which is tiny

Vector<StringData> strings; //resizes every powers of 2, so this is should already be more efficient and cache local than buckets

```